ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,969 @@
+import           GHC.Generics
+
+import           Control.Tracer (nullTracer)
+
+import           Cardano.Crypto.Hash.Class
+import           Cardano.Crypto.Hash.SHA256
+import           Cardano.Slotting.Slot
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.HeaderValidation
+import           Ouroboros.Consensus.Ledger.Basics
+import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Ledger.SupportsProtocol
+                     (LedgerSupportsProtocol)
+import           Ouroboros.Consensus.Mempool hiding (getTxSize)
+import           Ouroboros.Consensus.Mempool.Impl.Types (ForgeLedgerState (..),
+                     tickLedgerState)
+import           Ouroboros.Consensus.Mempool.TxSeq
+import           Ouroboros.Consensus.Storage.ChainDB.API hiding
+                     (getLedgerTablesAtFor)
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD.DiffSeq as DS
+import           Ouroboros.Consensus.Util
+import           Ouroboros.Consensus.Util.IOLike hiding (bracket)
+
+import           Test.QuickCheck
+import           Test.QuickCheck.Monadic
+import           Test.StateMachine hiding ((:>))
+import           Test.StateMachine.DotDrawing
+import qualified Test.StateMachine.Types as QC
+import           Test.StateMachine.Types (History (..), HistoryEvent (..))
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+
+import           Ouroboros.Consensus.Mock.Ledger.Address
+import           Ouroboros.Consensus.Mock.Ledger.Block
+import           Ouroboros.Consensus.Mock.Ledger.State
+import           Ouroboros.Consensus.Mock.Ledger.UTxO (Expiry, Tx, TxIn, TxOut)
+
+import           Test.Consensus.Mempool.Util (TestBlock, genTxs, testLedgerConfig, testInitLedger, genValidTxs, applyTxToLedger)
+
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+{-------------------------------------------------------------------------------
+  Datatypes
+-------------------------------------------------------------------------------}
+
+-- | Whether the tip is reachable or not. Isomorphic to Bool
+data IsReachable = IsReachable | IsNotReachable deriving (Generic, ToExpr, NoThunks)
+
+instance Arbitrary IsReachable where
+  arbitrary = bool IsReachable IsNotReachable <$> arbitrary
+
+isReachable :: IsReachable -> Bool
+isReachable IsReachable = True
+isReachable IsNotReachable = False
+
+infixl 9 /&&
+
+(/&&) :: IsReachable -> IsReachable -> IsReachable
+IsNotReachable /&& _ = IsNotReachable
+_ /&& IsNotReachable = IsNotReachable
+_ /&& _ = IsReachable
+
+-- | The model
+data Model blk r = Model {
+    -- | The current tip on the mempool
+    modelMempoolIntermediateState :: !(TickedLedgerState blk ValuesMK)
+
+    -- | The base state on top of which the mempool was last synced.
+    --
+    -- Used to make sure we don't generate a state with the same point but
+    -- different utxos.
+  , modelMempoolBaseState :: !(LedgerState blk ValuesMK)
```

Comment:
Maybe something like: 

```haskell

data LedgerDBContents l = OnlyTip l | TipPlusPrev l l 




Ltip ----------------Ltip'
    \
      \ Mtip 

===========================================

Ltip'

Ltip 
    \
      \ Mtip 

```

Reply:
One problematic case Javier explained to me:

```haskell
-- Assume current ledger state is l

ChangeLedger l' >> ChangeLedger l''

-- where tip (l') /= tip (l), and tip (l) == tip (l'')

```

Then the mempool could ask tables at `tip (l)` and can get a potentially different UTxO! This will produce an error when revalidating previously added transactions.

Reply:
One property of the ledger interface we should document is that:

```haskell
do 
    res  <- getLedgerTablesAtFor pt keys 
    res' <- getLedgerTablesAtFor pt keys 
    assert $ fromMaybe True $ (==) <$> res <*> res'
```